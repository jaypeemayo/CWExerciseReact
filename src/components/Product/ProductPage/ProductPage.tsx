import { createProduct, deleteProduct, getAllProducts, IProductGetParam, updateProduct } from "../helpers/ProductHelper"
import React, { useEffect } from 'react'
import { Table } from "../../../common/Table";
import { ProductRow } from "../ProductRow/ProductRow";
import Pagination from '@material-ui/lab/Pagination';
import { CreateEditProduct } from "../CreateEditProduct/CreateEditProduct";
import { MessageBox } from "../../../common/MessageBox";
import { IProduct } from "..";
import { IProductTableInfo } from ".";




export const ProductPage: React.FunctionComponent = () => {
  const defaultParam = {
    pageNumber: 1,
    pageSize: 5,
    columnToSort: "Name",
    sortDirection: 0,
  } as IProductGetParam;

  const defaulProducTableInfo = {
    numberOfPages: 0,
    products: []
  } as IProductTableInfo;

  const [productsTableInfo, setProductsTableInfo] = React.useState<IProductTableInfo>(defaulProducTableInfo);
  const [pageParam, setPageParam] = React.useState<IProductGetParam>(defaultParam);
  const [modal, setModal] = React.useState<any>();

  const getProducts = () => {
    getAllProducts(pageParam).then(o => {
      console.log(JSON.stringify(o));
      setProductsTableInfo(o)
    });
  }

  useEffect(() => {
    getProducts();
  }, [pageParam])

  const showMessageBox = (title: string, message: string) => {
    return setModal(<MessageBox title={title} buttonSet={<button onClick={() => { setModal(null) }}>Close</button>}>{message}</MessageBox>);
  }

  const showConfirmationBox = (id: number, title: string, message: string) => {
    return setModal(<MessageBox title={title} buttonSet={<><button onClick={() => { handleDelete(id); setModal(null); }}>Yes</button><button onClick={() => { setModal(null) }}>No</button></>}>{message}</MessageBox>);
  }

  const showCreateEditProduct = (productToEdit: IProduct) => {
    return setModal(<CreateEditProduct productToEdit={productToEdit} CloseCreateUpdate={() => { setModal(null); }} OnSavedProduct={handleSaveProduct} />);
  }

  const handleSaveProduct = (isCreating: boolean, product: IProduct) => {
    if (!isCreating) {
      updateProduct(product).then(o => {
        if (o) {
          getProducts();
          showMessageBox("Update Success", "Product Updated!");
        }
        else {
          showMessageBox("Update Failed", "Product Failed to Updated!");
        }
      });

    }
    else {
      createProduct(product).then(o => {

        if (o) {
          getProducts();
          showMessageBox("Create Success", "Product Created!");
        }
        else {
          showMessageBox("UpdCreatete Failed", "Product Failed to Create!");
        }
      });
    }
  }

  const handleDelete = (productID: number) => {
    deleteProduct(productID).then(o => {
      if (o) {
        getProducts();
      }
      else {
        showMessageBox("Delete Failed", "Product failed to delete!");
      }
    });
  }

  const handlePageChange = (event: React.ChangeEvent<unknown>, page: number) => {
    setPageParam({ ...pageParam, pageNumber: page })
  }

  const handleClickHeader = (columnToSort: string) => {
    setPageParam({ ...pageParam, pageNumber: 1, sortDirection: pageParam.sortDirection ? 0 : 1, columnToSort })
  }

  return (
    <div>
      <div>
        <button onClick={() => showCreateEditProduct(null)}>Create Product</button>
        <Table columnNames={["ProductID", "Name", "Price", "Type", "Active", ""]}
          sortedColumnName={pageParam.columnToSort}
          isAscending={pageParam.sortDirection ? true : false}
          onClickHeader={handleClickHeader}>
          <>
            {productsTableInfo.products.map((product, i) => <ProductRow OnEdit={showCreateEditProduct} OnDelete={(id) => { showConfirmationBox(id, "Delete Product", "Are you sure you want to delete the product?") }} key={i} {...product} />)}
          </>
        </Table>
        <Pagination count={productsTableInfo.numberOfPages} page={pageParam.pageNumber} shape="rounded" onChange={handlePageChange} />
      </div>
      {modal}

    </div>

  );
}
