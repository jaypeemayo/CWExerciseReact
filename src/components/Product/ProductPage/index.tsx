import { IProduct } from "..";

export interface IProductTableInfo {
  numberOfPages: number,
  products: IProduct[],
}
