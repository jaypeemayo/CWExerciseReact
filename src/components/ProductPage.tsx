import {getAllProducts, IProductGetParam} from "../helpers/ProductHelper"
import React, { useEffect } from 'react'
import { hot } from "react-hot-loader";
import { Table } from "../common/Table";
import { Product } from "./Product";
import Modal from 'react-modal';
import { CpuInfo } from "os";
import { FaChevronUp, FaChevronDown } from 'react-icons/fa';
import Pagination from '@material-ui/lab/Pagination';

const axios = require('axios');


export interface IProductPage {
  CustomerId: number; //to remove
}

export interface IProduct {
  productID: number;
  name: string;
  price: number;
  type: number;
  active: boolean;
}


export interface IProductTableInfo {
  numberOfPages : number,
  products : IProduct[],
}


export const ProductPage: React.FunctionComponent<IProductPage> = ({CustomerId }) => {

   const defaultParam = {
      pageNumber: 1,
      pageSize: 5,
      columnToSort: "Name",
      sortDirection: 1,
  } as IProductGetParam;

  const defaulProducTableInfo = {
    numberOfPages : 0,
    products: []
  } as IProductTableInfo;

  const  getProducts = () => {
    getAllProducts(pageParam).then(o=>{
      console.log(JSON.stringify(o));
         setProductsTableInfo(o)});
  }

  //const [customID, setCusetomerId] = React.useState<number>(CustomerId);
  const [productsTableInfo, setProductsTableInfo] = React.useState<IProductTableInfo>(defaulProducTableInfo);
  const [isCreating, setIsCreating] = React.useState<boolean>(false); //todo isCreatingUpdate
  const [pageParam, setPageParam] = React.useState<IProductGetParam>(defaultParam);

  useEffect(() => {
    getProducts();
  }, [pageParam])


  const createProduct = () =>{
    setIsCreating(true);
  }


  const closeCreateUpdate = () =>{
    setIsCreating(false);
  }

  const afterOpenModal = () =>{

  }
  const closeModal = () =>{
    
  }
  const sortRows = (columnToSort: string) =>{    
    setPageParam({...pageParam, pageNumber:1, sortDirection: pageParam.sortDirection ? 0 : 1, columnToSort})
  }

  const handlePageChange = (event: React.ChangeEvent<unknown>, page: number) =>{
    setPageParam({...pageParam, pageNumber: page})
  }

  const customStyles = {
    content: {
      top: '50%',
      left: '50%',
      right: 'auto',
      bottom: 'auto',
      marginRight: '-50%',
      transform: 'translate(-50%, -50%)',
    },
  };

  return (
    
  <div>
    {!isCreating ? 
    <div>
      <button onClick={createProduct}>Create Product</button>
      <Table columnNames = {["ProductID", "Name", "Price", "Type", "Active"]} 
        sortedColumnName={pageParam.columnToSort} 
        isAscending={pageParam.sortDirection ? true : false}
        onClickHeader={sortRows}>
      <>
        {productsTableInfo.products.map((product, i) => <Product key={i} {...product} />)}
      </>
      </Table>
      <Pagination count={productsTableInfo.numberOfPages} page={pageParam.pageNumber}  shape="rounded" onChange={handlePageChange} />
    </div>:
    <div>
      CREATING!
      <button onClick={closeCreateUpdate}>Close</button>
      <button onClick={closeCreateUpdate}>Save</button>
    </div>}

    <Modal
      isOpen={false}
      onAfterOpen={afterOpenModal}
      onRequestClose={closeModal}
      style={customStyles}
      contentLabel="Are you sure you want to Delete?"
    >
      Im a modal!
    </Modal>



    

  
  </div>



  );
}
