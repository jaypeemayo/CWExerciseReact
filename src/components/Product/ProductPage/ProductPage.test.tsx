import { render } from "@testing-library/react";
import * as React from "react";
import { IProduct } from "..";
import { ProductPage } from "./ProductPage";
var axios = require("axios");
var MockAdapter = require("axios-mock-adapter");

jest.mock("../../../common/Table/Table", () => {
    return {
        Table: (props: any) => {
            return <table>{props.children}</table>;
        }
    };
});

jest.mock("../ProductRow/ProductRow", () => {
    return {
        Table: (props: any) => {
            return <div>product row</div>;
        }
    };
});

jest.mock("@material-ui/lab/Pagination", () => ()=>{return <div>pagination mock</div>});

jest.mock("../CreateEditProduct/CreateEditProduct", () => {
    return {
        CreateEditProduct: (props: any) => {
            return <div>create edit product mock</div>;
        }
    };
});


jest.mock("../../../common/MessageBox/MessageBox", () => {
    return {
        MessageBox: (props: any) => {
            return <div>create edit product mock</div>;
        }
    };
});

test("Test snapshot", () => {
    var product = {
        productID: 1,
        name: "string",
        price: 2,
        type: 2,
        active: false,
    } as IProduct;
    const {container, unmount} = render(
        <ProductPage />
    );
    expect(container).toMatchSnapshot();
    unmount();
});


test("test load rows", () => {
    var mock = new MockAdapter(axios);
    mock.onGet("http://localhost:5000/api/product/?pageNumber=1&pageSize=5&columnToSort=Name&sortDirection=0").reply(200, {
        data: [product],
      });

    var product = {
        productID: 1,
        name: "string",
        price: 2,
        type: 2,
        active: false,
    } as IProduct;
    const {container, unmount} = render(
        <ProductPage />
    );
    expect(container).toMatchSnapshot();
    unmount();
});


