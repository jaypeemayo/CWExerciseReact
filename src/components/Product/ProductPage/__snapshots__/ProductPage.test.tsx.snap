// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <h5
        class="pt-5 pb-1"
      >
        Product Management System
      </h5>
      <table />
      <div
        class="d-flex justify-content-end"
      >
        <div>
          pagination mock
        </div>
      </div>
      <div
        class="d-flex justify-content-center"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          Create Product
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`test load rows 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <h5
        class="pt-5 pb-1"
      >
        Product Management System
      </h5>
      <table />
      <div
        class="d-flex justify-content-end"
      >
        <div>
          pagination mock
        </div>
      </div>
      <div
        class="d-flex justify-content-center"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          Create Product
        </button>
      </div>
    </div>
  </div>
</div>
`;
